### ISSUE TITLE
FloydWarshall is very memory inefficient

### ISSUE THREAD

**Issue by shevek** [jgrapht/jgrapht#561]
Further to our sending Dijkstra out to lunch:

FloydWarshall starts with:

        // init the backtrace matrix
        backtrace = new Object[n][n];

You know what's coming next :grinning: Let's assume that N = 586165 (our "small" graph):

```
586165 * 586165 * 8 // on a 64-bit machine
2748715257800 // bytes
./1048576
2621379 // megabytes...
./1024
2559 // gigabytes ...
```

Since our graph is directed and largely acyclic, I think we can do much better. Interestingly, we haven't even hit hash tables, this is a straight array. So, raising a ticket as a placeholder while we figure this out.

It looks like a lovely candidate for a ForkJoinPool, too.

**Comment by shevek** [comment]
Initial observations:
* The implementation of SingleSourcePaths can be reduced to:

```
    private class FloydWarshallSingleSourcePaths implements SingleSourcePaths<V, E> {
        private final V source;
        public FloydWarshallSingleSourcePaths(@Nonnull V source) {
            if (!graph.containsVertex(source))
                throw new IllegalArgumentException(GRAPH_MUST_CONTAIN_THE_SOURCE_VERTEX);
            this.source = source;
        }

        @Override
        public Graph<V, E> getGraph() {
            return graph;
        }

        @Override
        public V getSourceVertex() {
            return source;
        }

        @Override
        public double getWeight(V sink) {
            return FastFloydWarshallShortestPaths.this.getPathWeight(source, sink);
        }

        @Override
        public GraphPath<V, E> getPath(V sink) {
            return FastFloydWarshallShortestPaths.this.getPath(source, sink);
        }
    }
```

with no loss of efficiency - just a reduction in code size.

* There's an early termination condition in the cubic loop which is not exploited: If we go an entire iteration of `k` without adding an edge, break. This reduces the runtime from forgetaboutit to relatively-tractable.
* GnpGraphGenerator is inefficient in its edge-creation routine because it goes to the Map for `s` and `t` before it's evaluated `p`. I'm modelling our GNP graph on our real-world example, which has p=0.000738.

In any case, these various optimizations plus casting FloysWarshall in terms of a fastutil Long2DoubleMap (long = v_1 << 32 | v_2) (etc for the other substructures) has made the FloydWarshall tractable for sizes up to 5,000 fairly easily. I think 10K is right out, though. Oh well. Next plan.

**Comment by shevek** [comment]
I'm trying to think of other optimizations, for example, in the final loop, I think there's no point taking k to be any node with a degree of 1.

**Comment by jkinable** [comment]
What type of network are you trying to solve an all-pair shortest path problem on? Is it dense? I've successfully used JGraphT's implementation of Johnson's algorithm to solve an all-pair shortest path of a graph representing a city road network containing about 20k nodes. That I could solve a few minutes.

**Comment by shevek** [comment]
I don't actually need all-pairs, I just need pairs on-demand, and doing dijkstra on demand died badly in some cases. So I was thinking to do an all-pair up front, and then just read the table. All I can really tell you about it is that it's not-quite acyclic, N = 500_000 and p = 0.000738.

I note that in my test case (N=6_000, p=0.000738), the last loop terminated at k=24 not k=6000.
* If we sort the nodes by degree high to low before running, high-degree first, it takes k=(oh god i got bored and killed it)
* If we sort the nodes by degree low to high, it completes in k=0. Er. Wut?

But I think we have a clear answer. Sort nodes by degree, low to high, before running FloydWarshall. That wasn't my first hypothesis but now I know it's true, I can see why.

There HAS to be a PhD in this somewhere. Right?

**Comment by shevek** [comment]
No, I've done a miscalculation of p from our production dataset... I considered a subset of the nodes, but counted all the edges. I knew SOMETHING was wrong.

**Comment by shevek** [comment]
I've raised everything else in individual tickets, because they're relatively cheap, and this ticket really becomes a request to to:

1. Make the SingleSourcePaths call the methods in the outer class, rather than duplicating all the code.

And, subject to its availability:
1. Use a Long2DoubleOpenHashMap to store `d` 
1. Use a Long2ObjectOpenHashMap to store `backtrace`

at which point large graphs become tractable on my laptop.

**Comment by shevek** [comment]
I think we're going to have two issues with Johnson's:
* It is based on Dijkstra, and we already know that pairwise Dijkstra goes to lunch for a very long time on this graph.
* All our weights are 1, so we don't really need the Bellman-Ford step.
* The theory of FloydWarshall was that it's a form of simulated annealing, and with good pruning on `k` (per #563) should in practice be faster than a repeated Dijkstra even though it has a larger O-notation.

I'm having trouble visualizing the case in which FloydWarshall actually requires omega(n^3) steps. It seems that if all weights are positive, particularly if all weights are 1, it may be n^2.logn? In order to need N steps at k, you need two nodes separated by a minimum of N edges. But a straight-line graph of length N will solve in log(N) steps. But I'm sleep-deprived right now.

**Comment by d-michail** [comment]
We also have a bidirectional Dijkstra implementation where you provide both endpoints and does two searches until they meet. It is usually much faster and should help in your case. See `BidirectionalDijkstraShortestPath`.

**Comment by jsichi** [comment]
@shevek yes, we should clean up that duplicated code.

Sleep deprivation makes some of your leaps hard to follow (I'm not seeing how Floyd Warshall is a form of simulated annealing), but I'm sure there's a paper in there somewhere; the all pairs shortest paths problem is still an active research area, with sub-cubic bounds for unweighted, integer weights, etc.  

Our TransitiveClosure implementation makes use of the log(N) iteration capping you mention, but I don't see how that applies to Floyd Warshall.

**Comment by simlu** [comment]
@shevek Just stumbled across this and from my experience current Johnson implementation is much, much worse memory wise than Floyd. I'm just curious if you tried both and what you ended up going with!

My observations can be found here: https://github.com/jgrapht/jgrapht/issues/626

Looking forward to learning about your experience!

**Comment by shevek** [comment]
@simlu We did our own high-performance thing using fastutil. I strongly recommend the improvements in #563 which make the FloydWarshall a few orders of magnitude faster, but we got several orders of magnitude faster than that again with a fully custom implementation. If I ever get time, I would love to write a short paper on the performance improvements in FW, and why sorting works. Technically I owe you either a write-up or a patch regarding #563, because what I put in there was in note form and needs expanding.

Johnson's and bidirectional Dijkstra were both useless to us, far, far too slow and no obvious opportunity to improve them. Particularly the repeated Dijkstra has an outer linear loop which can't be reduced, and sucks fairly hard for memory locality (which itself carries a 200:1 performance penalty on x86).

**Comment by simlu** [comment]
@shevek Thank you very much for the reply!

What types of graphs are you dealing with? Sparse, Dense, degree distribution?

Your performance improvements sound pretty crazy - would you be able to plug in my [test case data](https://github.com/jgrapht/jgrapht/issues/626#issuecomment-404731029) and see what time you achieve? I don't want to go through a bunch of work only to find out that for my graphs the improvement is not significant or even negative.

For my graphs the difference between floyd and johnson (jgrapht implementations) is pretty significant (103 seconds vs 6 seconds for my test). I was under the impression that johnson is superior unless the graph is dense? So I'm assuming your graphs are dense?

**Comment by jsichi** [comment]
(I think the duplicated code noted by @shevek has finally been removed in #630.)

**Comment by jsichi** [comment]
I've merged #632.  Remaining items here are:

* fastutil variant
* Alternative all-pairs algorithms for special cases (e.g. integer edges) where sub-cubic bounds are known; these should be addressed in their own issues.
* Cases where the outer loop can terminate early; I do not believe this is applicable in general, but this may be relevant for some of the special cases (like in TransitiveClosure).

Anything else?
